### ISSUE TITLE
LTeX doesn't recognize macron

### ISSUE THREAD

**Issue by intractabilis** [valentjn/vscode-ltex#269]
Usually, if I have a word unfamiliar to LTeX, it underlines it and in the tooltip I can choose to add the word to dictionary. If the word has letters with grave or accute accent, or with circumflex, LTeX properly adds all diacritics to the dictionary. However, it skips letters with macrons completely.

Takea look at the following MWE:
```latex
\documentclass[11pt, paper = B5]{scrbook}
\usepackage{fontspec}
\setmainfont{Libertinus Serif}
\begin{document}
Nih\'oŋ\=go
\end{document}
```
It's a single word "Nihóŋḡo". Here is how the tooltip looks like:

![изображение](https://user-images.githubusercontent.com/45959195/111023611-da89f100-838e-11eb-9eb1-c3e665fbc014.png)

You can see it is "Nihóŋo", instead of "Nihóŋḡo". "Nihóŋo" as well will be added to dictionary.

- Operating system: Linux (linux), x64, 5.10.19-1-MANJARO
- VS Code: 1.53.2
- vscode-ltex: 9.0.0
- ltex-ls: 10.0.0
- Java: 15.0.2

**Comment by valentjn** [comment]
Thanks for the report. LT<sub>E</sub>X needs to [manually map each letter to the correct Unicode code point](https://github.com/valentjn/ltex-ls/blob/a407c2726c092f9569cc8c66be6a3abedb727c87/ltexls-core/src/main/java/org/bsplines/ltexls/parsing/latex/LatexAnnotatedTextBuilder.java#L965-L979). As `g` isn't in there, it will be skipped. It would probably already be a first step if it fell back to the unaccented letter instead of just skipping it altogether, despite still being incorrect.

Do you just need support for `g`? Or do you also need other letters? If yes, which ones?

**Comment by intractabilis** [comment]
I am typesetting a text with romanization of Japanese pronunciation. Acute, grave and circumflex are used in that romanization to indicate pitch variation and can happen above any letter, even consonant. For example:

![изображение](https://user-images.githubusercontent.com/45959195/111919173-d6706a00-8a45-11eb-90c6-5c3c3a1ed347.png)

What would be ideal is if I could add `itte` and `kara` without diacritics to LTeX dictionary and it could skip them as correct in the text regardless of the diacritics.

Macron is used only above n̄ and ḡ to indicate nasal pronunciation. But n̄ with macron can also have acute above macron:

![изображение](https://user-images.githubusercontent.com/45959195/111919512-9ad69f80-8a47-11eb-9b7b-b453e3d64b36.png)

In this case it would be great to have **ben̄kyoo** in the dictionary.

**Comment by valentjn** [comment]
Ah, okay. Then we'll have to use Unicode's combining diacritical marks as there is a Unicode character `ḡ` with macron for `g`, but of course not for every letter, and even less for two or more diacritics at the same time. I can add support for this, but of course LanguageTool has to support it as well, which it probably does. As a workaround, if your L<sup>A</sup>T<sub>E</sub>X engine supports Unicode (e.g., LuaL<sup>A</sup>T<sub>E</sub>X), you should already be able to use Unicode instead of `\'`, `\=`, etc. in your documents.

I think we can't just ignore the diacritics, otherwise we'd have many false negatives for all the languages which use them (languages in which omitting them would be a spelling error). So in your last example, it would add `ben̄́kyoo` or `ben̄́kyoo-site` to the dictionary (not just `ben̄kyoo`), depending on what LanguageTool underlines.

**Comment by intractabilis** [comment]
> As a workaround, if your LATEX engine supports Unicode (e.g., LuaLATEX), you should already be able to use Unicode instead of `\'`, `\=`, etc. in your documents.

That doesn't make any difference, ḡ is still ignored:

![изображение](https://user-images.githubusercontent.com/45959195/112692062-8e818680-8e3b-11eb-8d0c-141b7808bc30.png)

> I think we can't just ignore the diacritics, otherwise we'd have many false negatives for all the languages which use them (languages in which omitting them would be a spelling error). So in your last example, it would add `ben̄́kyoo` or `ben̄́kyoo-site` to the dictionary (not just `ben̄kyoo`), depending on what LanguageTool underlines.

It's absolutely fine. If it just doesn't ignore letters with diacritics at all (like ḡ above), it's all good. Thank you!

**Comment by valentjn** [comment]
Edit: I noticed that Unicode places the combining diacritical marks *after* the letter, not before like in this comment. The point is still valid, though.

That doesn't make any sense, there's no logic in LT<sub>E</sub>X that would remove this, in contrast to `\=`. It's working for me for both single-character Unicode (U+1E21: `ḡ`) and Unicode with combining diacritical marks on ASCII characters (U+0304 U+0067: `ḡ `):

![image](https://user-images.githubusercontent.com/19839841/112719425-3dd06300-8ef9-11eb-9f00-e74cf73a21f4.png)

Hex dump of the file:

```plaintext
00000000  4e 69 68 c3 b3 c5 8b e1  b8 a1 6f 0a 0a 4e 69 68  |Nih.......o..Nih|
00000010  6f cc 81 c5 8b 67 cc 84  6f 0a                    |o....g..o.|
```

**Comment by intractabilis** [comment]
Sorry, mea culpa. You are right. My LTeX is set to check on save, and I forgot to save when I experimented. Thanks for the workaround!

**Comment by ed359** [comment]
See also https://github.com/valentjn/ltex-ls/issues/56#issue-847269345. Having \H{o} and \H{u} would be useful in Hungarian. I made a PR https://github.com/valentjn/ltex-ls/pull/57

**Comment by valentjn** [comment]
I replaced the whole accent table with Java's Unicode normalization algorithm. This means that combined characters will be used if they exist, otherwise combining diacritical marks will be used ([Normalization Form C in Unicode Standard Annex #15](https://www.unicode.org/reports/tr15/tr15-50.html#Normalization_Forms_Table)).

Combining multiple accents on a single letter (e.g., `\'{\=n}`) is still not supported, since L<sup>A</sup>T<sub>E</sub>X itself seems to have difficulties with this without special packages. Users that rely on multiple accents on a single letter need to use Unicode instead.

**Comment by valentjn** [comment]
Fix released in 10.0.0.

**Comment by intractabilis** [comment]
Thanks!
